File: rgbd_data_adapter/src/depth_to_laser_adapter.py
#! /usr/bin/env python
#Ubuntu 16.04, ROS Kinetic, python 2.7
import rospy
import rosbag
from sensor_msgs.msg   import LaserScan
from tf2_msgs.msg      import TFMessage 
from geometry_msgs.msg import TransformStamped
import tf
from tf.transformations import euler_from_quaternion
#Node that reads data published from node pointcloud_to_laserscan and rewrites data with correct topic name and framed to be accepted by the AMCL node

#Get Launch File Parameters
TF_x      = rospy.get_param('/depth_to_laser_adapter/TF_x')
TF_y      = rospy.get_param('/depth_to_laser_adapter/TF_y')
TF_z      = rospy.get_param('/depth_to_laser_adapter/TF_z')
TF_yaw    = rospy.get_param('/depth_to_laser_adapter/TF_yaw')
rgbd_id = rospy.get_param('/depth_to_laser_adapter/rgbd_id')
pc_to_laser_topic = rospy.get_param('/depth_to_laser_adapter/pc_to_laser_topic')

pub_scan = rospy.Publisher('depth'+str(rgbd_id), LaserScan, queue_size = 11)
pub_tf   = rospy.Publisher("tf", TFMessage, queue_size=11)   
scan_received = LaserScan()    

def create_TFmsg(x, y, z, yaw, frame, child_frame, seq, t):
    trans = TransformStamped()
    trans.header.seq = seq
    trans.header.stamp = t
    trans.header.frame_id = frame
    trans.child_frame_id = child_frame
    trans.transform.translation.x = x
    trans.transform.translation.y = y
    trans.transform.translation.z = z
    q = tf.transformations.quaternion_from_euler(0,0,yaw)
    trans.transform.rotation.x = q[0]
    trans.transform.rotation.y = q[1]
    trans.transform.rotation.z = q[2]
    trans.transform.rotation.w = q[3]
    msg = TFMessage()
    msg.transforms.append(trans)
    return msg

def callback(msg):
    current_time = rospy.Time(0)
    scan_received.header.seq         = msg.header.seq 
    scan_received.header.stamp     = current_time
    scan_received.header.frame_id = 'depth'+str(rgbd_id)
    scan_received.angle_min           = msg.angle_min      
    scan_received.angle_max          = msg.angle_max      
    scan_received.angle_increment= msg.angle_increment
    scan_received.time_increment   = msg.time_increment 
    scan_received.scan_time            = msg.scan_time      
    scan_received.range_min           = msg.range_min      
    scan_received.range_max          = msg.range_max      
    scan_received.ranges                  = msg.ranges         
    
    new_tf_msg   = create_TFmsg(TF_x, TF_y, TF_z, TF_yaw, 'base_link','depth'+str(rgbd_id), msg.header.seq+1, current_time)
    pub_scan.publish(scan_received)
    pub_tf.publish(new_tf_msg)

def listener():
    rospy.init_node('depth_to_laser_adapter', anonymous=True)
    sub = rospy.Subscriber(pc_to_laser_topic, LaserScan, callback)
    rospy.spin()

if __name__ == '__main__':
    listener()
 

File: rgbd_data_adapter/src/depth_to_laser_adapter.cpp
//Ubuntu 20.04LTS, ROS Noetic
//Node that reads data published from node pointcloud_to_laserscan and rewrites data with correct topic name and framed to be accepted by the AMCL node
#include<ros/ros.h>
#include<std_msgs/Int32.h>
#include<sensor_msgs/LaserScan.h>
#include<tf2_msgs/TFMessage.h>
#include<geometry_msgs/TransformStamped.h>
#include<tf/transform_datatypes.h>
#include<tf/transform_broadcaster.h>
#include<string>

//Global Variables
sensor_msgs::LaserScan scan_received;
int rgbd_id, TF_yaw;
float TF_x, TF_y, TF_z;
std::string pc_to_laser_topic = "depth";

void scanCallback(const sensor_msgs::LaserScan::ConstPtr &msg) {
    scan_received.header.seq             = msg->header.seq; 
    scan_received.header.stamp         = msg->header.stamp;
    scan_received.angle_min              = msg->angle_min;      
    scan_received.angle_max             = msg->angle_max;      
    scan_received.angle_increment   = msg->angle_increment;
    scan_received.time_increment      = msg->time_increment; 
    scan_received.scan_time               = msg->scan_time;      
    scan_received.range_min              = msg->range_min;      
    scan_received.range_max             = msg->range_max;      
    scan_received.ranges                     = msg->ranges;         
    }
 
//MAIN PROGRAM
int main(int argc, char **argv)
{
	//Initialize
	ros::init(argc, argv, "depth_to_laser_adapter");
	ros::NodeHandle nh;
	
	//Declare publishers and suscriber
	ros::Subscriber sub            = nh.subscribe(pc_to_laser_topic, 1000, scanCallback);
	ros::Publisher  pub_scan  = nh.advertise<sensor_msgs::LaserScan>("laser", 1000); 
	//ros::Publisher  pub_tf         = nh.advertise<tf2_msgs::TFMessage>("tf",1000);
	tf::TransformBroadcaster lasertf_broadcaster;
	
	//Get params from launch file
	nh.getParam("/depth_to_laser_adapter/TF_x",TF_x);
	nh.getParam("/depth_to_laser_adapter/TF_y",TF_y);
	nh.getParam("/depth_to_laser_adapter/TF_z",TF_z);
	nh.getParam("/depth_to_laser_adapter/TF_yaw",TF_yaw);
	nh.getParam("/depth_to_laser_adapter/rgbd_id",rgbd_id);
	nh.getParam("/depth_to_laser_adapter/pc_to_laser_topic",pc_to_laser_topic);
	
	//Node frequency
	ros::Rate r(10);
	
	while(nh.ok()){
		//current_time = ros::Time::now(); //////////////////////////////////////////get current time
		//publish laser msg over ROS
	    sensor_msgs::LaserScan laser_msg;
		laser_msg.header.frame_id = "laser";
		laser_msg.header.seq         = scan_received.header.seq;
		laser_msg.header.stamp     = scan_received.header.stamp;
		laser_msg.angle_min           =scan_received.angle_min;      
		laser_msg.angle_max          =scan_received.angle_max;      
		laser_msg.angle_increment=scan_received.angle_increment;
		laser_msg.time_increment   =scan_received.time_increment; 
		laser_msg.scan_time            =scan_received.scan_time;      
		laser_msg.range_min           =scan_received.range_min;      
		laser_msg.range_max          =scan_received.range_max;      
		laser_msg.ranges                  =scan_received.ranges;    
		//publish the msg
		pub_scan.publish(laser_msg);
		
		////transform theta to quaternion for odom msg, odom_quat
		//geometry_msgs::Quaternion lasertf_quat = tf::createQuaternionMsgFromYaw(TF_yaw);
		////publishh transform over tf
		//geometry_msgs::TransformStamped laser_tf;
		//laser_tf.header.seq                = scan_received.header.seq;
		//laser_tf.header.stamp            = scan_received.header.stamp;
		//laser_tf.header.frame_id       = "laser";
		//laser_tf.child_frame_id          = "base_link";
		//laser_tf.transform.translation.x = TF_x;
		//laser_tf.transform.translation.y = TF_y;
		//laser_tf.transform.translation.z = TF_z;
		//laser_tf.transform.rotation         = lasertf_quat;
		////send transform
		//lasertf_broadcaster.sendTransform(laser_tf);	
		
		static tf::TransformBroadcaster br;
		tf::Transform transform;
		transform.setOrigin( tf::Vector3(TF_x, TF_y, TF_z) );
		transform.setRotation( tf::Quaternion(TF_yaw, 0, 0) );
		br.sendTransform(tf::StampedTransform(transform, scan_received.header.stamp, "base_link", "laser"));

		ros::spinOnce();  //check for incoming messages
		r.sleep();
	}
	//ros::spin();
	//return 0;
}

